### PR TITLE
Minify combined class names

### PR DESCRIPTION
**Pull request by gilbox**
Hashes the combined class name to create a minified combined class name.

@lencioni
@xymostech
@jlfwong

### CODE REVIEW

File: src/exports.js [hunk 1]
```diff
@@ -21,7 +21,6 @@ const StyleSheet = {
     create(sheetDefinition /* : SheetDefinition */) {
         return mapObj(sheetDefinition, ([key, val]) => {
             return [key, {
-                // TODO(gil): Further minify the -O_o--combined hashes
                 _name: process.env.NODE_ENV === 'production' ?
                     `_${hashObject(val)}` : `${key}_${hashObject(val)}`,
```
Comment (lencioni) [inline src/exports.js]: Since this is going to be joined and hashed again, we can probably remove the leading underscore here now, right?

**Comment by gilbox** [comment]
ping @lencioni addressed your comment and added other improvements

**Comment by gilbox** [comment]
ping @xymostech

**Comment by csilvers** [comment]
If you need more than 32 bits of hash, you're kinda stuck, since js doesn't provide 64-bit integer types.

You could go back to using a google hash: https://github.com/lovell/farmhash e.g.  It has fingerprint64() which gives you a 64-bit hash (as a string), and is platform independent.  I don't know how fast it is, though.

Your only other solution would be to generate two separate 32-bit hashes and concatenate them.  The most straightforward way to do this, if you're always hashing the same kind of object, is to add a 'salt' field to the object, and hash it once with salt=1 and a second time with salt=2 (or whatever).

Of course, this will double your hashing time.  farmhash is probably faster. :-)

**Comment by csilvers** [comment]
(Blah, maybe not possible -- it looks like farmhash is node-specific, and just a wrapper around some C code.)

File: src/inject.js [hunk 2]
```diff
@@ -269,7 +269,15 @@ export const injectAndGetClassName = (
     if (processedStyleDefinitions.classNameBits.length === 0) {
         return "";
     }
-    const className = processedStyleDefinitions.classNameBits.join("-o_O-");
+
+    let className;
+    if (process.env.NODE_ENV === 'production') {
+        className = processedStyleDefinitions.classNameBits.length === 1 ?
+            `_${processedStyleDefinitions.classNameBits[0]}` :
+            `_${hashString(processedStyleDefinitions.classNameBits.join())}`;
```
Comment (lencioni) [inline src/inject.js]: What if we append `processedStyleDefinitions.classNameBits.length`? That would help reduce the likelihood of collisions.

**Comment by gilbox** [comment]
ping @xymostech @lencioni 
I implemented something similar to @lencioni's suggestion but instead of using the length of `processedStyleDefinitions.classNameBits.length` I used the combined length of the stringified style definition objects since they probably exhibit more variation. However, that length is modded by `36` so that there is only one additional byte added.

File: src/inject.js [hunk 3]
```diff
@@ -243,6 +243,12 @@ const processStyleDefinitions = (
     }
 };
 
+// Sum up the lengths of the stringified style definitions (which was saved as _len property)
+// and use modulus to return a single byte hash value.
```
Comment (lencioni) [inline src/inject.js]: This comment would be 💯  if it explained why we are adding the length to the hash.

File: src/inject.js [hunk 4]
```diff
@@ -269,7 +275,16 @@ export const injectAndGetClassName = (
     if (processedStyleDefinitions.classNameBits.length === 0) {
         return "";
     }
-    const className = processedStyleDefinitions.classNameBits.join("-o_O-");
+
+    let className;
+    if (process.env.NODE_ENV === 'production') {
+        className = processedStyleDefinitions.classNameBits.length === 1 ?
+            `_${processedStyleDefinitions.classNameBits[0]}` :
```
Comment (xymostech) [inline src/inject.js]: should we maybe be adding the length of the style definitions when there's only one class name, too? If only to make the hashes the same size/consistent?

**Comment by xymostech** [comment]
I think the travis failing was a fluke, I restarted it

File: src/inject.js (hunk 4 above)
Comment (gilbox) [inline src/inject.js]: I'm not sure what the advantage of that would be. Also, before this PR the hashes were different lengths--for example, from our homepage:

```
image_ay4wjb-o_O-background_1h6n1zu-o_O-fadeIn_3jddj2-o_O-backgroundSize_cover_176vplr
```
--so I imagine they still won't be the same size.

Comment (gilbox) [inline src/inject.js]: oh, the other thing is that we'd be hashing a hash unnecessarily. is the additional expense justified?

File: src/inject.js (hunk 3 above)
Comment (gilbox) [inline src/inject.js]: https://github.com/Khan/aphrodite/pull/248/commits/263b35f98e512a8af345d02bd34d9894e0e3c655

File: src/inject.js (hunk 4 above)
Comment (xymostech) [inline src/inject.js]: We wouldn't be hashing anything new, just adding the extra length char on, right? You're right though, this probably isn't important. :)

Comment (gilbox) [inline src/inject.js]: oh, you're right double-hashing shouldn't be a thing

**Comment by xymostech** [comment]
Okay, I'm gonna merge this! Sorry for the delay! Maybe once this lands, @lencioni or I could try updating aphrodite on our sites and see if anything breaks and how much smaller the styles get.

**Comment by lencioni** [comment]
That sounds good. I actually work with @gilbox so I'll let him do the honors on our site and report back here with anything interesting. 💃

**Comment by xymostech** [comment]
Oh ha, I didn't realize! Sounds wonderful :)

**Comment by gilbox** [comment]
@xymostech could we get a release please 🙏

**Comment by xymostech** [comment]
@gilbox Did you test things out on your site? I can make a release but I'd love for someone to do some real testing to make sure this doesn't horribly break things...

**Comment by gilbox** [comment]
ah ok I didn't realize that's what you meant. I'll give it a whirl

**Comment by gilbox** [comment]
I tested with #263, should be good

**Comment by gilbox** [comment]
ping @xymostech for the new version

**Pull request merged**